config/load.go

```diff
@@ -0,0 +1,298 @@
+		if flag, ok := flags[value.Pointer()]; ok {
+			cfg.Log.Trace(fmt.Sprintf("binding: %s = %v (flag)\n", strings.ToUpper(regexp.MustCompile("[^a-zA-Z0-9]").ReplaceAllString(appPrefix+path, "_")), value.Elem().Interface()))
+			_ = v.BindPFlag(path, flag)
+			return
+		}
+	}
+
+	if value.Type().Kind() == reflect.Ptr {
+		value = value.Elem()
+	}
+
+	if value.Type().Kind() != reflect.Struct {
+		cfg.Log.Trace(fmt.Sprintf("binding: %s = %v\n", strings.ToUpper(regexp.MustCompile("[^a-zA-Z0-9]").ReplaceAllString(appPrefix+path, "_")), value.Interface()))
+		v.SetDefault(path, value.Interface())
+		return
+	}
+
+	if path != "" {
+		path += "."
+	}
+
+	// for each field in the configuration struct, see if the field implements the defaultValueLoader interface and invoke it if it does
+	for i := 0; i < value.NumField(); i++ {
+		fieldValue := value.Field(i)
+		field := value.Type().Field(i)
+
+		mapStructTag := field.Tag.Get("mapstructure")
+
+		if mapStructTag == "-" {
+			continue
+		}
+
+		if !field.Anonymous && mapStructTag == "" {
+			cfg.Log.Trace(fmt.Sprintf("not binding field due to lacking mapstructure tag: %s.%s", value.Type().Name(), field.Name))
+			continue
+		}
+
+		if fieldValue.Type().Kind() != reflect.Ptr {
+			fieldValue = fieldValue.Addr()
+		}
+
+		configureViper(cfg, v, fieldValue, flags, appPrefix, path+mapStructTag)
+	}
+}
+
+type flagRefs map[uintptr]*pflag.Flag
+
+func getFlagRefs(cmd *cobra.Command) flagRefs {
+	refs := flagRefs{}
+	for _, flags := range []*pflag.FlagSet{cmd.PersistentFlags(), cmd.Flags()} {
+		flags.VisitAll(func(flag *pflag.Flag) {
+			v := reflect.ValueOf(flag.Value)
+			// check for struct types like stringArrayValue
+			if v.Type().Kind() == reflect.Ptr {
+				vf := v.Elem()
+				if vf.Type().Kind() == reflect.Struct {
+					if _, ok := vf.Type().FieldByName("value"); ok {
+						vf = vf.FieldByName("value")
+						if vf.IsValid() {
+							v = vf
+						}
+					}
+				}
+			}
+			refs[v.Pointer()] = flag
+		})
+	}
+	return refs
+}
+
+//nolint:unused
+func hasConfig(base string) bool {
+	for _, ext := range viper.SupportedExts {
+		if _, err := os.Stat(fmt.Sprintf("%s.%s", base, ext)); err != nil {
+			return true
+		}
+	}
+	return false
+}
+
+// nolint:funlen
+func loadConfig(cfg Config, v *viper.Viper, appName string, configPath string) error {
+	var err error
+	// use explicitly the given user config
+	if configPath != "" {
+		v.SetConfigFile(configPath)
+		if err := v.ReadInConfig(); err != nil {
+			return fmt.Errorf("unable to read application config=%q : %w", configPath, err)
+		}
+		v.Set("config", v.ConfigFileUsed())
+		// don't fall through to other options if the config path was explicitly provided
+		return nil
+	}
+
+	// start searching for valid configs in order...
+	// 1. look for .<appname>.yaml (in the current directory)
+	confFilePath := "." + appName
+
+	// TODO: Remove this before v1.0.0
+	// See syft #1634
+	v.AddConfigPath(".")
+	v.SetConfigName(confFilePath)
```

Comment:
Should we rewrite this function to do the accurate loading behavior or continue to have it attempting to load the wrong file(s)?

Reply:
actually, we could have each individual loading behavior as separate functions? (I could branch off of this to illustrate) then the called could collect which loading behaviors they want (including for syft to use the old/buggy loading option).